File: lab5/SparkLab5Task1/src/main/java/com/github/rmdarth/bdpclab5/MonthAccumulator.java
package com.github.rmdarth.bdpclab5;

import org.apache.spark.util.AccumulatorV2;
import scala.Tuple2;

import java.util.*;

public class MonthAccumulator extends AccumulatorV2<Tuple2<Integer, Long>, List<Long>>  {
    private List<Long> monthData;

    MonthAccumulator() {
        monthData = new ArrayList<>(Collections.nCopies(12, 0L));
    }

    MonthAccumulator(List<Long> data) {
        monthData = data;
    }

    @Override
    public boolean isZero() {
        for (int i = 0; i < 12; i++)
            if (monthData.get(i) != 0)
                return false;
        return true;
    }

    @Override
    public AccumulatorV2 copy() {
        return new MonthAccumulator(new ArrayList<>(monthData));
    }

    @Override
    public void reset() {
        for (int i = 0; i < 12; i++)
            monthData.set(i, 0L);
    }

    @Override
    public void add(Tuple2<Integer, Long> v) {
        monthData.set(v._1 - 1, monthData.get(v._1 - 1) + v._2);
    }

    @Override
    public void merge(AccumulatorV2<Tuple2<Integer, Long>, List<Long>> other) {
        for (int i = 0; i < 12; i++)
            monthData.set(i, monthData.get(i) + other.value().get(i));
    }

    @Override
    public List<Long> value() {
        return monthData;
    }
}

File: lab6/SparkDF1/src/main/scala/com/github/rmdarth/bdpclab6/MonthAccumulator.scala
package com.github.rmdarth.bdpclab6

import org.apache.spark.util.AccumulatorV2

import scala.collection.mutable

// Accumulator containing 12 counters for each month
class MonthAccumulator extends AccumulatorV2[(Int, Long), mutable.Seq[Long]] {
  var monthData: mutable.Seq[Long] = mutable.Seq.fill(12)(0L)

  def this(newList: mutable.Seq[Long]) = {
    this()
    monthData = newList.clone()
  }

  override def isZero: Boolean = {
    for (i <- 0 until 12)
      if (monthData(i) != 0L) return false
    true
  }
  override def copy() = new MonthAccumulator(monthData)
  override def reset(): Unit = monthData = mutable.Seq.fill(12)(0L)
  override def add(data: (Int, Long)) = {
    monthData(data._1 - 1) = monthData(data._1 - 1) + data._2
  }
  override def merge(other: AccumulatorV2[(Int, Long), mutable.Seq[Long]]): Unit = {
    for (i <- 0 until 12)
      monthData(i) = monthData(i) + other.value(i)
  }
  override def value: mutable.Seq[Long] = monthData
}
